doc/macro_utils_design.md

```diff
@@ -0,0 +1,240 @@
+        1) `__VA_ARGS__ ()` expanded to have a comma or
+        2) `MU_TRIGGER_PARENTHESIS __VA_ARGS__` expanded to have a comma. 
+    In both cases "1" and "2" the macro is not empty. (these are not the droids we're looking for).
+
+The combination of `MU_HAS_COMMA` + `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS __VA_ARGS__ ())` + `MU_HAS_COMMA(__VA_ARGS__ ())` + `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS __VA_ARGS__)` are all set in a dispath table from which the final result of `MU_ISEMPTY` is deduced.
+
+Here are some examples of how this works:
+                   `MU_HAS_COMMA`               `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS __VA_ARGS__ ())`           `MU_HAS_COMMA(__VA_ARGS__ ())`         `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS __VA_ARGS__)`
+
+MU_ISEMPTY() = >        0                       `MU_HAS_COMMA((MU_TRIGGER_PARENTHESIS())`                       `MU_HAS_COMMA(())`                     `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS)`
+                        0                       `MU_HAS_COMMA(,)                                                0                                       0
+                        0                       1                                                               0                                       0
+
+MU_ISEMPTY(A)=>         0                       `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS A ())`                     `MU_HAS_COMMA(A())`                     `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS A)`
+                        0                       0                                                               0                                        0
+
+#define M6() , /*note here, only M6 followed by () expands to ",", otherwise M6 stays as it is
+MU_ISEMPTY(M6)=>        0                       `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS M6 ())`                    `MU_HAS_COMMA(M6())`                    `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS M6)`
+                        0                       `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS ,    )`                    `MU_HAS_COMMA(,)`                       `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS M6)`
+                        0                       1                                                               1                                       0
+
+MU_ISEMPTY(() Z)=>      0                       `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS () Z)`                     `MU_HAS_COMMA(() Z())`                  `MU_HAS_COMMA(MU_TRIGGER_PARENTHESIS () Z)`
+                        0                       `MU_HAS_COMMA( , Z )`                                           `MU_HAS_COMMA(() Z())`                  `MU_HAS_COMMA(, Z)`
+                        0                       1                                                               0                                       1
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
```

Comment:
Lots of empty lines that can be removed

Reply:
Removed

---
In reply to: [464138072](https://github.com/Azure/azure-macro-utils-c/pull/42#discussion_r464138072) [](ancestors = 464138072)